File: server/src/service.ts
import type {
  ChapterFile,
  Meta,
  Progress,
  IndexEntry,
  StoryStatus,
} from "@ao3hub/shared";
import { Meta as MetaSchema } from "@ao3hub/shared";
import { story, upsertIndex, patchIndex, loadIndex } from "./db";
import { parseAo3Html } from "./ao3/parse";
import { fetchDownloadHtml, extractWorkId } from "./ao3/fetch";
import { makeBlankTranslated, runTranslation } from "./translate/worker";
import { enqueue, setProcessor } from "./queue";
import { emit } from "./sse";

function nowIso(): string {
  return new Date().toISOString();
}

function randomId(): string {
  return "u" + Math.random().toString(36).slice(2, 10) + Date.now().toString(36).slice(-4);
}

function indexEntryFor(meta: Meta, status: StoryStatus): IndexEntry {
  return {
    id: meta.id,
    title: meta.title,
    chineseTitle: meta.chineseTitle,
    author: meta.author,
    chapterCount: meta.chapterCount,
    wordCount: meta.wordCount,
    status,
    addedAt: nowIso(),
    updatedAt: nowIso(),
  };
}

async function persistParsed(
  html: string,
  source: { url?: string; downloadUrl?: string; workId?: string },
): Promise<{ meta: Meta; original: ChapterFile; isNew: boolean }> {
  const parsed = parseAo3Html(html);
  const id = source.workId ?? parsed.meta.workIdGuess ?? randomId();
  const url = source.url ?? parsed.meta.workUrlGuess ?? `https://archiveofourown.org/works/${id}`;
  const meta: Meta = MetaSchema.parse({
    ...parsed.meta,
    id,
    url,
    downloadUrl: source.downloadUrl,
  });

  const isNew = !(await story.exists(id));
  await story.saveSource(id, html);
  await story.saveMeta(id, meta);
  await story.saveOriginal(id, parsed.original);

  const existing = await story.loadTranslated(id);
  const translated =
    existing && existing.chapters.length === parsed.original.chapters.length
      ? existing
      : makeBlankTranslated(parsed.original);
  await story.saveTranslated(id, translated);

  const total = parsed.original.chapters.reduce(
    (acc, c) => acc + c.blocks.filter((b) => b.type !== "hr" && b.html.trim()).length,
    0,
  );
  const done = translated.chapters.reduce(
    (acc, c) => acc + c.blocks.filter((b) => b.status === "done").length,
    0,
  );
  const progress: Progress = {
    phase: "queued",
    totalBlocks: total,
    doneBlocks: done,
    startedAt: nowIso(),
    errors: [],
  };
  await story.saveProgress(id, progress);

  return { meta, original: parsed.original, isNew };
}

export async function createFromUrl(url: string): Promise<{ id: string; status: StoryStatus }> {
  const workId = extractWorkId(url);
  if (!workId) throw new Error("无法从 URL 提取 work id");

  await prepareEntry({ id: workId, title: "Fetching…", author: "", status: "fetching" });
  emit(workId, { type: "phase", phase: "fetching" });

  let html: string;
  try {
    html = await fetchDownloadHtml(workId);
  } catch (e) {
    await patchIndex(workId, { status: "error" });
    emit(workId, { type: "phase", phase: "error", message: e instanceof Error ? e.message : String(e) });
    throw e;
  }

  const { meta } = await persistParsed(html, {
    url,
    downloadUrl: `https://archiveofourown.org/works/${workId}?view_full_work=true`,
    workId,
  });
  await upsertIndex(indexEntryFor(meta, "queued"));
  enqueue({ storyId: workId, type: "translate" });
  return { id: workId, status: "queued" };
}

export async function createFromHtml(html: string): Promise<{ id: string; status: StoryStatus }> {
  const { meta } = await persistParsed(html, {});
  await upsertIndex(indexEntryFor(meta, "queued"));
  enqueue({ storyId: meta.id, type: "translate" });
  return { id: meta.id, status: "queued" };
}

async function prepareEntry(partial: {
  id: string;
  title: string;
  author: string;
  status: StoryStatus;
}): Promise<void> {
  const idx = await loadIndex();
  const existing = idx.stories.find((s) => s.id === partial.id);
  if (existing) {
    await patchIndex(partial.id, { status: partial.status });
    return;
  }
  await upsertIndex({
    id: partial.id,
    title: partial.title,
    author: partial.author,
    chapterCount: 0,
    wordCount: 0,
    status: partial.status,
    addedAt: nowIso(),
    updatedAt: nowIso(),
  });
}

export async function retryStory(
  id: string,
  opts: { blockIds?: string[]; chapterIndex?: number },
): Promise<void> {
  const translated = await story.loadTranslated(id);
  const original = await story.loadOriginal(id);
  if (!translated || !original) throw new Error("story not found");

  const idSet = opts.blockIds ? new Set(opts.blockIds) : null;
  for (let ci = 0; ci < translated.chapters.length; ci++) {
    if (typeof opts.chapterIndex === "number" && ci !== opts.chapterIndex) continue;
    const ch = translated.chapters[ci];
    for (let bi = 0; bi < ch.blocks.length; bi++) {
      const b = ch.blocks[bi];
      if (idSet && !idSet.has(b.id)) continue;
      if (!idSet && b.status !== "error") continue;
      const ob = original.chapters[ci].blocks[bi];
      if (ob.type === "hr" || !ob.html.trim()) continue;
      ch.blocks[bi] = { ...b, status: "pending", html: "", error: undefined };
    }
  }
  await story.saveTranslated(id, translated);
  await patchIndex(id, { status: "translating" });
  enqueue({ storyId: id, type: "translate" });
}

export async function deleteStory(id: string): Promise<void> {
  const idx = await loadIndex();
  idx.stories = idx.stories.filter((s) => s.id !== id);
  await (await import("./db")).saveIndex(idx);
  await story.remove(id);
}

export function initWorker(): void {
  setProcessor(async (job) => {
    try {
      if (job.type === "translate" || job.type === "retry") {
        await runTranslation(job.storyId);
      }
    } catch (e) {
      console.error(`[worker] story ${job.storyId} failed:`, e);
      const msg = e instanceof Error ? e.message : String(e);
      await patchIndex(job.storyId, { status: "error" });
      emit(job.storyId, { type: "phase", phase: "error", message: msg });
    }
  });
}

export async function resumeOnStartup(): Promise<void> {
  const idx = await loadIndex();
  for (const entry of idx.stories) {
    if (entry.status === "ready" || entry.status === "error") continue;
    const p = await story.loadProgress(entry.id);
    if (!p) continue;
    if (p.phase === "ready" || p.phase === "error") continue;
    enqueue({ storyId: entry.id, type: "translate" });
  }
}

File: internal/app/service.go
package app

import (
	"errors"
	"fmt"
	"strings"
)

func indexEntryFor(meta Meta, status StoryStatus) IndexEntry {
	now := nowISO()
	return IndexEntry{
		ID:           meta.ID,
		Title:        meta.Title,
		ChineseTitle: meta.ChineseTitle,
		Author:       meta.Author,
		ChapterCount: meta.ChapterCount,
		WordCount:    meta.WordCount,
		Status:       status,
		AddedAt:      now,
		UpdatedAt:    now,
	}
}

func (a *App) persistParsed(html string, source struct {
	URL         string
	DownloadURL string
	WorkID      string
}, mode TranslationMode) (Meta, ChapterFile, bool, error) {
	parsed, err := parseAO3HTML(html)
	if err != nil {
		return Meta{}, ChapterFile{}, false, err
	}
	id := source.WorkID
	if id == "" {
		id = parsed.Meta.WorkIDGuess
	}
	if id == "" {
		id = randomStoryID()
	}
	url := source.URL
	if url == "" {
		url = parsed.Meta.WorkURLGuess
	}
	if url == "" {
		url = fmt.Sprintf("https://archiveofourown.org/works/%s", id)
	}
	meta := parsed.Meta.Meta
	meta.ID = id
	meta.URL = url
	meta.DownloadURL = source.DownloadURL

	existing, _ := a.store.LoadMeta(id)
	switch {
	case strings.TrimSpace(string(mode)) != "":
		meta.TranslationMode = normalizeTranslationMode(mode)
	case existing != nil:
		meta.TranslationMode = existing.TranslationMode
	}
	meta = normalizeMeta(meta)

	isNew := !a.store.StoryExists(id)
	if err := a.store.SaveSource(id, html); err != nil {
		return Meta{}, ChapterFile{}, false, err
	}
	if err := a.store.SaveMeta(id, meta); err != nil {
		return Meta{}, ChapterFile{}, false, err
	}
	if err := a.store.SaveOriginal(id, parsed.Original); err != nil {
		return Meta{}, ChapterFile{}, false, err
	}

	translated, _ := a.store.LoadTranslated(id)
	var nextTranslated ChapterFile
	if translatedMatchesOriginal(translated, parsed.Original) {
		nextTranslated = *translated
	} else {
		nextTranslated = makeBlankTranslated(parsed.Original)
	}
	if err := a.store.SaveTranslated(id, nextTranslated); err != nil {
		return Meta{}, ChapterFile{}, false, err
	}

	total := 0
	for _, chapter := range parsed.Original.Chapters {
		for _, block := range chapter.Blocks {
			if isTranslatable(block) {
				total++
			}
		}
	}
	done := 0
	for ci, chapter := range parsed.Original.Chapters {
		for bi, block := range chapter.Blocks {
			if !isTranslatable(block) || ci >= len(nextTranslated.Chapters) || bi >= len(nextTranslated.Chapters[ci].Blocks) {
				continue
			}
			if nextTranslated.Chapters[ci].Blocks[bi].Status == BlockDone {
				done++
			}
		}
	}
	progress := Progress{
		Phase:       PhaseQueued,
		TotalBlocks: total,
		DoneBlocks:  done,
		StartedAt:   nowISO(),
		Errors:      []ProgressError{},
	}
	if err := a.store.SaveProgress(id, progress); err != nil {
		return Meta{}, ChapterFile{}, false, err
	}

	return meta, parsed.Original, isNew, nil
}

func translatedMatchesOriginal(translated *ChapterFile, original ChapterFile) bool {
	if translated == nil || len(translated.Chapters) != len(original.Chapters) {
		return false
	}
	for ci, chapter := range original.Chapters {
		if len(translated.Chapters[ci].Blocks) != len(chapter.Blocks) {
			return false
		}
		for bi, block := range chapter.Blocks {
			tBlock := translated.Chapters[ci].Blocks[bi]
			if tBlock.ID != block.ID || tBlock.Type != block.Type {
				return false
			}
		}
	}
	return true
}

func (a *App) prepareEntry(id, title, author string, status StoryStatus) error {
	idx, err := a.store.LoadIndex()
	if err != nil {
		return err
	}
	for _, entry := range idx.Stories {
		if entry.ID == id {
			_, err := a.store.PatchIndex(id, func(e *IndexEntry) {
				e.Status = status
			})
			return err
		}
	}
	return a.store.UpsertIndex(IndexEntry{
		ID:           id,
		Title:        title,
		Author:       author,
		ChapterCount: 0,
		WordCount:    0,
		Status:       status,
		AddedAt:      nowISO(),
		UpdatedAt:    nowISO(),
	})
}

func (a *App) CreateFromURL(url string, mode TranslationMode) (map[string]string, error) {
	workID := extractWorkID(url)
	if workID == "" {
		return nil, errors.New("无法从 URL 提取 work id")
	}
	if err := a.prepareEntry(workID, "Fetching…", "", StatusFetching); err != nil {
		return nil, err
	}
	a.bus.Emit(workID, StreamEvent{Type: "phase", Phase: PhaseFetching})

	html, err := a.fetchDownloadHTML(workID)
	if err != nil {
		_, _ = a.store.PatchIndex(workID, func(entry *IndexEntry) {
			entry.Status = StatusError
		})
		a.bus.Emit(workID, StreamEvent{Type: "phase", Phase: PhaseError, Message: err.Error()})
		return nil, err
	}

	meta, _, _, err := a.persistParsed(html, struct {
		URL         string
		DownloadURL string
		WorkID      string
	}{
		URL:         url,
		DownloadURL: fmt.Sprintf("https://archiveofourown.org/works/%s?view_full_work=true", workID),
		WorkID:      workID,
	}, mode)
	if err != nil {
		return nil, err
	}
	if err := a.store.UpsertIndex(indexEntryFor(meta, StatusQueued)); err != nil {
		return nil, err
	}
	a.queue.Enqueue(Job{StoryID: workID, Type: "translate"})
	return map[string]string{"id": workID, "status": string(StatusQueued)}, nil
}

func (a *App) CreateFromHTML(html string, mode TranslationMode) (map[string]string, error) {
	meta, _, _, err := a.persistParsed(html, struct {
		URL         string
		DownloadURL string
		WorkID      string
	}{}, mode)
	if err != nil {
		return nil, err
	}
	if err := a.store.UpsertIndex(indexEntryFor(meta, StatusQueued)); err != nil {
		return nil, err
	}
	a.queue.Enqueue(Job{StoryID: meta.ID, Type: "translate"})
	return map[string]string{"id": meta.ID, "status": string(StatusQueued)}, nil
}

func (a *App) RetryStory(id string, blockIDs []string, chapterIndex *int, mode TranslationMode) error {
	translated, err := a.store.LoadTranslated(id)
	if err != nil || translated == nil {
		return errors.New("story not found")
	}
	original, err := a.store.LoadOriginal(id)
	if err != nil || original == nil {
		return errors.New("story not found")
	}
	if strings.TrimSpace(string(mode)) != "" {
		nextMode := normalizeTranslationMode(mode)
		if meta, _ := a.store.LoadMeta(id); meta != nil && meta.TranslationMode != nextMode {
			meta.TranslationMode = nextMode
			if err := a.store.SaveMeta(id, *meta); err != nil {
				return err
			}
		}
	}
	idSet := map[string]bool{}
	if len(blockIDs) > 0 {
		for _, id := range blockIDs {
			idSet[id] = true
		}
	}
	for ci := range translated.Chapters {
		if chapterIndex != nil && ci != *chapterIndex {
			continue
		}
		for bi := range translated.Chapters[ci].Blocks {
			block := translated.Chapters[ci].Blocks[bi]
			if len(idSet) > 0 && !idSet[block.ID] {
				continue
			}
			if len(idSet) == 0 && block.Status != BlockError {
				continue
			}
			if ci >= len(original.Chapters) || bi >= len(original.Chapters[ci].Blocks) {
				continue
			}
			ob := original.Chapters[ci].Blocks[bi]
			if !isTranslatable(ob) {
				block.Status = BlockDone
				block.HTML = ob.HTML
				block.Error = ""
				translated.Chapters[ci].Blocks[bi] = block
				continue
			}
			block.Status = BlockPending
			block.HTML = ""
			block.Error = ""
			translated.Chapters[ci].Blocks[bi] = block
		}
	}
	if err := a.store.SaveTranslated(id, *translated); err != nil {
		return err
	}
	if err := a.updateStoryStatus(id, StatusTranslating); err != nil {
		return err
	}
	a.queue.Enqueue(Job{StoryID: id, Type: "translate"})
	return nil
}

func (a *App) DeleteStory(id string) error {
	idx, err := a.store.LoadIndex()
	if err != nil {
		return err
	}
	next := idx.Stories[:0]
	for _, entry := range idx.Stories {
		if entry.ID != id {
			next = append(next, entry)
		}
	}
	idx.Stories = next
	if err := a.store.SaveIndex(idx); err != nil {
		return err
	}
	return a.store.RemoveStory(id)
}

func (a *App) ResumeOnStartup() error {
	idx, err := a.store.LoadIndex()
	if err != nil {
		return err
	}
	for _, entry := range idx.Stories {
		if entry.Status == StatusReady || entry.Status == StatusError {
			continue
		}
		progress, err := a.store.LoadProgress(entry.ID)
		if err != nil || progress == nil {
			continue
		}
		if progress.Phase == PhaseReady || progress.Phase == PhaseError {
			continue
		}
		a.queue.Enqueue(Job{StoryID: entry.ID, Type: "translate"})
	}
	return nil
}
